### PR TITLE
Dynamically render credential subform fields

### PR DESCRIPTION
**Pull request by mabashian**
##### SUMMARY
Link #5879 

One of the first decisions I made was to actually not dynamically generate the _whole_ form but just the type details subform area.  All credentials have a Name/Description/Organization field so I felt like it was OK to leave that area of the form alone and just focus on the subform.  There are a number of new/updated components and I'm going to try to run through them and explain what's going on in each.

- `CredentialForm.jsx`: A couple of big changes in here.  Firstly, I found it easier to go ahead and initialize all the possible field values right off the bat rather than trying to do this on the fly as the credential type changed.  We'll see corresponding logic in the add/edit components that filters values down to only fields that belong to the selected credential type before sending the POST/PUT.  You'll also see an extension of the reset logic.  I tried to line this behavior up with the old application.  The basic gist is: if you change credential types it resets all the subform fields to their default value as provided by the API.  The one exception to this is when you switch back to the saved credential type (on edit).  In this case we'll revert the fields back to their _saved_ values.  The last change in here has to do with the subform - instead of having a big conditional where we include different subform components based on credential type we now render a `CredentialSubForm` component that handles the rendering of all of the relevant fields (when a cred type is selected).
- `CredentialSubForm.jsx`: This splits the fields into two buckets (bool & non-bool) then loops across each of those arrays.  For bool fields, they all get lumped together into a single OPTIONS `FormGroup`.  For non-bool fields, a `CredentialField` is rendered.
- `CredentialField.jsx`: This component introspects the field options and decides how it should be rendered.  For external cred types (like CyberArk and Hashi) we don't want to render the credential plugin fields to we render those as normal `FormGroup`'s.  If the field has `choices` then we know it should be a dropdown and render it appropriately.  By default, most fields are going to be rendering the `CredentialPluginField`.
- `CredentialPluginField.jsx`: This component handles all the plugin related logic and wraps the input in an InputGroup in order to add that `key` button.
- `BecomeMethodField.jsx`: This field needed special treatment because we maintain the array of default options on the client (not available via the API).
- `GceFileUploadField.jsx`: This is not actually a field but rather a shortcut users can use to populate the rest of the GCE subform.
- `CredentialAdd.jsx/CredentialEdit.jsx`: The big change here has to do with bundling up the payload that we send to `/api/v2/credentials`.  We now use the credential type OPTIONS response to filter out input fields that don't apply to the selected credential type.  There's also added logic for handling `Prompt on launch` inputs.  This differs from `Prompt on launch` fields in the JT form in that they don't have their own field on the model.  So, to enable prompting on a field we have to pass the `ASK` string in the payload.

**Outstanding issues (things that are not included in this PR):**
- `Replace` functionality on password fields: https://github.com/ansible/awx/issues/7256
- Update credential details to handle plugin input fields: https://github.com/ansible/awx/issues/7127
- Hook up credential plugin `Test` button: https://github.com/ansible/awx/issues/7126
- Drag and drop on key fields: https://github.com/ansible/awx/issues/6703
- Add file picker support for key fields: https://github.com/ansible/awx/issues/6113

Once we're settled on the approach I'll certainly add more test coverage to the form(s) but wanted to avoid unnecessary work if we decided to go in a different direction.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

### CODE REVIEW

File: awx/ui_next/src/screens/Credential/shared/CredentialSubForm.jsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import React from 'react';
```
Comment (jlmitch5) [inline awx/ui_next/src/screens/Credential/shared/CredentialSubForm.jsx]: maybe we should name this file CredentialTypeSubForm?

File: awx/ui_next/src/screens/Credential/CredentialAdd/CredentialAdd.jsx [hunk 2]
```diff
@@ -44,18 +59,17 @@ function CredentialAdd({ me }) {
           inputs: nonPluginInputs,
           ...remainingValues,
         });
-        const inputSourceRequests = [];
-        Object.entries(pluginInputs).forEach(([key, value]) => {
-          inputSourceRequests.push(
+
+        await Promise.all(
```
Comment (jlmitch5) [inline awx/ui_next/src/screens/Credential/CredentialAdd/CredentialAdd.jsx:64]: at first i was thinking it'd be better to concat plugin/non-plugin together, but I think the way useCallback and await works...it'll still finish all this before returning that it's finished

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx [hunk 3]
```diff
@@ -3,54 +3,73 @@ import { Formik, useField } from 'formik';
...
+    .map(key => {
+      return {
+        value: credentialTypes[key].id,
+        key: credentialTypes[key].id,
+        label: credentialTypes[key].name,
+      };
+    })
+    .sort((a, b) => (a.label > b.label ? 1 : -1));
+
+  const resetSubFormFields = (newCredentialType, form) => {
```
Comment (jlmitch5) [inline awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx:37]: would this be better in "CredentialSubForm"?

File: awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+function CredentialField({ credentialType, fieldOptions, i18n }) {
+  const { values: formikValues } = useFormikContext();
+  const requiredFields = credentialType?.inputs?.required || [];
+  const isRequired = requiredFields.includes(fieldOptions.id);
+  const [subFormField, meta, helpers] = useField({
+    name: `inputs.${fieldOptions.id}`,
+    validate:
+      isRequired && !formikValues?.passwordPrompts[fieldOptions.id]
```
Comment (jlmitch5) [inline awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx]: can we unroll this from the ternaries and && into a function?

**Review by jlmitch5 — COMMENTED** [review]
this is looking great!  just a couple things hear and there that might could be tweak, but i love the direction!

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx [hunk 5]
```diff
@@ -3,54 +3,73 @@ import { Formik, useField } from 'formik';
...
+    credentialTypes[newCredentialType].inputs.fields.forEach(
+      ({ ask_at_runtime, type, id, choices, default: defaultValue }) => {
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx:40]: I think formik's resetForm function can help with some of the complicated logic here. check this out https://github.com/ansible/awx/blob/devel/awx/ui_next/src/screens/Inventory/shared/InventorySourceForm.jsx#L28

If I had my way I would get the api to tell me whether the cred types are custom or not when we fetch them from the api.  Then, when in edit mode, I would only show those types in the drop down, and if there are no custom creds, the drop down is disabled.

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx [hunk 6]
```diff
@@ -135,80 +147,54 @@ function CredentialForm({
...
-      ssh_key_unlock: credential?.inputs?.ssh_key_unlock || '',
-      ssh_public_key_data: credential?.inputs?.ssh_public_key_data || '',
-      username: credential?.inputs?.username || '',
-    },
+    inputs: {},
+    passwordPrompts: {},
   };
 
+  Object.values(credentialTypes).forEach(credentialType => {
+    credentialType.inputs.fields.forEach(
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx]: I created a custom cred type in the old ui that did not have any input configuration or injector configuration and it broke this line of code.  I realize that a user creating such a Cred Type is unlikely, however, we should be prepared for it.

File: awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx (hunk 4 above)
Comment (AlexSCorey) [reply awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx]: +1

**Review by AlexSCorey — CHANGES_REQUESTED** [review]
- [ ] We definitely need to figure out the bug mentioned below with a custom cred that has no input configuration.  

- [ ] I think we also need to re-think those nested ternaries as they can be hard to follow.

- [ ] As far as the reset function goes, I think I can be cleaned up a bit, especially the first if statement and I'd be happy to chat with you further about that if necessary.

File: awx/ui_next/src/screens/Credential/shared/CredentialSubForm.jsx (hunk 1 above)
Comment (mabashian) [reply awx/ui_next/src/screens/Credential/shared/CredentialSubForm.jsx]: After chatting with @jlmitch5 about this one I think we've settled on `TypeInputsSubForm`.  If anyone else has ideas for this component name then let us know.

File: awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx (hunk 4 above)
Comment (mabashian) [reply awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx]: Sure, np

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx (hunk 5 above)
Comment (jakemcdermott) [reply awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx:40]: Credential types have a `managed_by_tower` field. `false` means custom.

I think this drop-down would show _both_ custom and non-custom credentials, though. For example, I'd expect to be able to select the machine credential type or gce type even though both are managed (non-custom).

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx (hunk 6 above)
Comment (mabashian) [reply awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx]: 👍 will def fix this

File: awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx (hunk 4 above)
Comment (mabashian) [reply awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx]: Talked with @jlmitch5 about this one and I think we came up with more readable solution.

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx [hunk 7]
```diff
@@ -34,45 +29,6 @@ function CredentialFormFields({
     })
     .sort((a, b) => (a.label > b.label ? 1 : -1));
 
-  const resetSubFormFields = (newCredentialType, form) => {
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx]: https://media.giphy.com/media/Is1O1TWV0LEJi/giphy.gif

File: awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx [hunk 8]
```diff
@@ -120,8 +76,16 @@ function CredentialFormFields({
             ...credentialTypeOptions,
           ]}
           onChange={(event, value) => {
-            credTypeHelpers.setValue(value);
-            resetSubFormFields(value, formik);
+            const { values, initialValues, resetForm } = formik;
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Credential/shared/CredentialForm.jsx]: wow, awesome!

File: awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx [hunk 9]
```diff
@@ -76,19 +76,20 @@ function CredentialField({ credentialType, fieldOptions, i18n }) {
   const { values: formikValues } = useFormikContext();
   const requiredFields = credentialType?.inputs?.required || [];
   const isRequired = requiredFields.includes(fieldOptions.id);
+  const validateField = () => {
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Credential/shared/CredentialFormFields/CredentialField.jsx:79]: 👍

**Review by AlexSCorey — APPROVED** [review]
I think this looks really good.

**Review by jlmitch5 — APPROVED** [review]
great stuff!

**Review by jakemcdermott — APPROVED** [review]

**Comment by dsesami** [comment]
@mabashian what do you want the general rules to be for this for form persistence? it's super weird currently, so I'd like to get some defined behavior down.

1. Make some AWS creds, but don't save. Switch to something else and then switch back to AWS.
- The AWS creds are cleared out.

2. Make some cyberark creds and then save. Go back and edit.
- If you switch back to AWS, your previously existing fields are there! Well, not all of them. I had to go back and forth saving a cyberark and aws cred to do this, but things which i suspect use the same "key" are persisted between forms when editing, like "username".

**Comment by dsesami** [comment]
followed up with @mabashian. Once we iron out the persistence in forms mentioned above + the weird issue of "Vault" appearing after the alphabetically-latter "VMware" this should be good for the purposes of this PR.

**Review by dsesami — APPROVED** [review]
barring the formik stuff I discussed with @mabashian which is in a different PR, this is good to go.
